core/src/impl/Kokkos_Profiling.cpp

```diff
@@ -0,0 +1,689 @@
+ // PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ // PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ // LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ // NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ // SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ //
+ // Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+ //
+ // ************************************************************************
+ //@HEADER
+ */
+
+#include <Kokkos_Macros.hpp>
+#include <dlfcn.h>
+#include <vector>
+#include <array>
+#if defined(KOKKOS_ENABLE_PROFILING)
+#include <impl/Kokkos_Profiling.hpp>
+#include <cstring>
+#include <unordered_map>
+#include <unordered_set>
+#include <algorithm>
+namespace Kokkos {
+namespace Profiling {
+
+static initFunction initProfileLibrary         = nullptr;
+static finalizeFunction finalizeProfileLibrary = nullptr;
+
+static beginFunction beginForCallback    = nullptr;
+static beginFunction beginScanCallback   = nullptr;
+static beginFunction beginReduceCallback = nullptr;
+static endFunction endForCallback        = nullptr;
+static endFunction endScanCallback       = nullptr;
+static endFunction endReduceCallback     = nullptr;
+
+static pushFunction pushRegionCallback = nullptr;
+static popFunction popRegionCallback   = nullptr;
+
+static allocateDataFunction allocateDataCallback     = nullptr;
+static deallocateDataFunction deallocateDataCallback = nullptr;
+
+static beginDeepCopyFunction beginDeepCopyCallback = nullptr;
+static endDeepCopyFunction endDeepCopyCallback     = nullptr;
+
+static createProfileSectionFunction createSectionCallback   = nullptr;
+static startProfileSectionFunction startSectionCallback     = nullptr;
+static stopProfileSectionFunction stopSectionCallback       = nullptr;
+static destroyProfileSectionFunction destroySectionCallback = nullptr;
+
+static profileEventFunction profileEventCallback = nullptr;
+}  // end namespace Profiling
+
+namespace Tuning {
+size_t getNewContextId();
+size_t getCurrentContextId();
+void decrementCurrentContextId();
+size_t getNewVariableId();
+
+static size_t kernel_name_context_variable_id;
+static size_t kernel_type_context_variable_id;
+static tuningVariableDeclarationFunction tuningVariableDeclarationCallback =
+    nullptr;
+static tuningVariableValueFunction tuningVariableValueCallback = nullptr;
+static contextVariableDeclarationFunction contextVariableDeclarationCallback =
+    nullptr;
+static contextEndFunction contextEndCallback = nullptr;
+
+}  // end namespace Tuning
+
+namespace Profiling {
+
+bool profileLibraryLoaded() { return (initProfileLibrary != nullptr); }
+
+void beginParallelFor(const std::string& kernelPrefix, const uint32_t devID,
+                      uint64_t* kernelID) {
+  if (beginForCallback != nullptr) {
+    Kokkos::fence();
+    (*beginForCallback)(kernelPrefix.c_str(), devID, kernelID);
+#ifdef KOKKOS_ENABLE_TUNING
```

Comment:
Maybe not appropriate for this PR, but it seems like the begin functions are replicating a lot of code? Would it be better to have a `beginParallel` and `endParallel` functions. Then `beginParallelFor` would just be:
````
void beginParallelFor(....){
  beginParallel("for", beginForCallback, kernPrefix, devID, kernelID);
}
````

Reply:
I get the idea here, but I don't know that it buys us enough on the tools side to be worth perturbing an interface that is supposed to be fairly stable. What I might do is on the tools side, have a utility library that does roughly the same thing.